In [1]:
## Standard Imports
import pandas as pd
## Importing the OS and JSON Modules
import os,json

In [61]:
df1 = pd.read_csv('superhero_info - superhero_info.csv')
df1.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [62]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


In [63]:
df1.columns

Index(['Hero|Publisher', 'Gender', 'Race', 'Alignment', 'Hair color',
       'Eye color', 'Skin color', 'Measurements'],
      dtype='object')

In [64]:
df1['Hero|Publisher'].head()

0            A-Bomb|Marvel Comics
1    Abe Sapien|Dark Horse Comics
2              Abin Sur|DC Comics
3       Abomination|Marvel Comics
4     Absorbing Man|Marvel Comics
Name: Hero|Publisher, dtype: object

In [65]:
df1['Hero|Publisher'].str.split("|", expand=True)

0                  1
0             A-Bomb      Marvel Comics
1         Abe Sapien  Dark Horse Comics
2           Abin Sur          DC Comics
3        Abomination      Marvel Comics
4      Absorbing Man      Marvel Comics
..               ...                ...
458     Yellowjacket      Marvel Comics
459  Yellowjacket II      Marvel Comics
460             Yoda       George Lucas
461          Zatanna          DC Comics
462             Zoom          DC Comics

[463 rows x 2 columns]

In [66]:
df1[['Hero','Publisher']] = df1['Hero|Publisher'].str.split("|", expand=True)
df1.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [67]:
df1 = df1.drop(columns='Hero|Publisher')
df1.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher  
0      Marvel Comics  
1  Dark Horse Comics  
2          DC Comics  
3      Marvel Comics  
4      Marvel Comics

In [68]:
measurements_df = df1.loc[0,"Measurements"]
print(type(measurements_df))
measurements_df

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [69]:
measurements_df = measurements_df.replace("'", '"')
measurements_df

'{"Height": "203.0 cm", "Weight": "441.0 kg"}'

In [70]:
fixed_m = json.loads(measurements_df)
print(type(fixed_m))
fixed_m

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [71]:
df1['Measurements'] = df1['Measurements'].str.replace("'", '"')
df1['Measurements'] = df1['Measurements'].apply(json.loads)

In [72]:
df1['Measurements'].head()

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
2     {'Height': '185.0 cm', 'Weight': '90.0 kg'}
3    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
4    {'Height': '193.0 cm', 'Weight': '122.0 kg'}
Name: Measurements, dtype: object

In [73]:
height_Weight = df1['Measurements'].apply(pd.Series)
height_Weight.head()

Height    Weight
0  203.0 cm  441.0 kg
1  191.0 cm   65.0 kg
2  185.0 cm   90.0 kg
3  203.0 cm  441.0 kg
4  193.0 cm  122.0 kg

In [74]:
df1 = pd.concat((df1, height_Weight), axis=1)
df1.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

                                   Measurements           Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}         A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}     Abe Sapien   
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}       Abin Sur   
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}    Abomination   
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}  Absorbing Man   

           Publisher    Height    Weight  
0      Marvel Comics  203.0 cm  441.0 kg  
1  Dark Horse Comics  191.0 cm   65.0 kg  
2          DC Comics  185.0 cm   90.0 kg  
3      Marvel Comics  203.0 cm  441.0 kg  
4      Marvel Comics  193.0 cm  122.0 kg

In [75]:
df1 = df1.drop(columns='Measurements')
df1.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher    Height    Weight  
0         A-Bomb      Marvel Comics  203.0 cm  441.0 kg  
1     Abe Sapien  Dark Horse Comics  191.0 cm   65.0 kg  
2       Abin Sur          DC Comics  185.0 cm   90.0 kg  
3    Abomination      Marvel Comics  203.0 cm  441.0 kg  
4  Absorbing Man      Marvel Comics  193.0 cm  122.0 kg

In [76]:
to_replace = ['cm', ' ']

for char in to_replace:
    df1['Height'] = df1['Height'].str.replace(char, "", regex=False)

df1['Height'].head()

0    203.0
1    191.0
2    185.0
3    203.0
4    193.0
Name: Height, dtype: object

In [77]:
df1['Height'] = pd.Series(df1['Height'], dtype=float)
df1['Height'].head()

0    203.0
1    191.0
2    185.0
3    203.0
4    193.0
Name: Height, dtype: float64

In [78]:
to_replace = ['kg', ' ']

for char in to_replace:
    df1['Weight'] = df1['Weight'].str.replace(char, "", regex=False)

df1['Weight'].head()

0    441.0
1     65.0
2     90.0
3    441.0
4    122.0
Name: Weight, dtype: object

In [79]:
df1['Weight'] = pd.Series(df1['Weight'], dtype=float)
df1['Weight'].head()

0    441.0
1     65.0
2     90.0
3    441.0
4    122.0
Name: Weight, dtype: float64

In [81]:
df1.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher  Height  Weight  
0         A-Bomb      Marvel Comics   203.0   441.0  
1     Abe Sapien  Dark Horse Comics   191.0    65.0  
2       Abin Sur          DC Comics   185.0    90.0  
3    Abomination      Marvel Comics   203.0   441.0  
4  Absorbing Man      Marvel Comics   193.0   122.0

In [145]:
df2 = pd.read_csv('superhero_powers - superhero_powers.csv')
df2.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [146]:
# split string on comma
df2['powers_split'] = df2['Powers'].str.split(',')

# check
df2[['Powers', 'powers_split']].head()

Powers  \
0         Agility,Super Strength,Stamina,Super Speed   
1  Accelerated Healing,Durability,Longevity,Super...   
2  Agility,Accelerated Healing,Cold Resistance,Du...   
3                                 Lantern Power Ring   
4  Accelerated Healing,Intelligence,Super Strengt...   

                                        powers_split  
0    [Agility, Super Strength, Stamina, Super Speed]  
1  [Accelerated Healing, Durability, Longevity, S...  
2  [Agility, Accelerated Healing, Cold Resistance...  
3                               [Lantern Power Ring]  
4  [Accelerated Healing, Intelligence, Super Stre...

In [147]:
exploded = df2.explode('powers_split')
exploded[['hero_names', 'Powers', 'powers_split']].head()

hero_names                                             Powers  \
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   

          powers_split  
0              Agility  
0       Super Strength  
0              Stamina  
0          Super Speed  
1  Accelerated Healing

In [148]:
cols_to_make = exploded['powers_split'].dropna().unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [149]:
frames = [df2]

for col in cols_to_make:
    frame = pd.DataFrame()
    frame[col] = df2['Powers'].str.contains(col)
    frames.append(frame)
    merged_powers = pd.concat(frames, axis = 1)
    
# check
merged_powers.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        powers_split  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   
3                               [Lantern Power Ring]    False           False   
4  [Accelerated Healing, Intelligence, Super Stre...    False            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   
3    False        False                False       False      False  ...   
4     True         True                 True       False      False  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 170 columns]

In [150]:
# drop 'Powers' and 'powers_split' columns
merged_powers.drop(columns = ['Powers', 'powers_split'], inplace = True)

# check
merged_powers.head()

hero_names  Agility  Super Strength  Stamina  Super Speed  \
0      3-D Man     True            True     True         True   
1       A-Bomb    False            True     True        False   
2   Abe Sapien     True            True     True        False   
3     Abin Sur    False           False    False        False   
4  Abomination    False            True     True         True   

   Accelerated Healing  Durability  Longevity  Camouflage  Self-Sustenance  \
0                False       False      False       False            False   
1                 True        True       True        True             True   
2                 True        True       True       False            False   
3                False       False      False       False            False   
4                 True       False      False       False            False   

   ...  Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0  ...            False        False       False              False   
1  ...            False        False       False              False   
2  ...            False        False       False              False   
3  ...            False        False       False              False   
4  ...            False        False       False              False   

   Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0       False       False          False               False    False   
1       False       False          False               False    False   
2       False       False          False               False    False   
3       False       False          False               False    False   
4       False       False          False               False    False   

   Changing Armor  
0           False  
1           False  
2           False  
3           False  
4           False  

[5 rows x 168 columns]

In [151]:
# change 'hero_names' column name to match info['Hero']
merged_powers.rename(columns = {'hero_names': 'Hero'}, 
                     inplace = True)

# check
merged_powers.head(2)

Hero  Agility  Super Strength  Stamina  Super Speed  \
0  3-D Man     True            True     True         True   
1   A-Bomb    False            True     True        False   

   Accelerated Healing  Durability  Longevity  Camouflage  Self-Sustenance  \
0                False       False      False       False            False   
1                 True        True       True        True             True   

   ...  Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0  ...            False        False       False              False   
1  ...            False        False       False              False   

   Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0       False       False          False               False    False   
1       False       False          False               False    False   

   Changing Armor  
0           False  
1           False  

[2 rows x 168 columns]

In [152]:
# check info for merged_powers before merging
merged_powers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Columns: 168 entries, Hero to Changing Armor
dtypes: bool(167), object(1)
memory usage: 114.1+ KB


In [153]:
# check info for info before merging
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Gender      463 non-null    object 
 1   Race        463 non-null    object 
 2   Alignment   463 non-null    object 
 3   Hair color  463 non-null    object 
 4   Eye color   463 non-null    object 
 5   Skin color  463 non-null    object 
 6   Hero        463 non-null    object 
 7   Publisher   463 non-null    object 
 8   Height      463 non-null    float64
 9   Weight      463 non-null    float64
dtypes: float64(2), object(8)
memory usage: 36.3+ KB


In [154]:
# combine two dataframes on info['Hero'] and merged_powers['Hero']
hero_df = pd.merge(df1, 
                   merged_powers, 
                   on = 'Hero')

# check
hero_df.head()

Gender               Race Alignment Hair color Eye color Skin color  \
0   Male              Human      good    No Hair    yellow    Unknown   
1   Male      Icthyo Sapien      good    No Hair      blue       blue   
2   Male            Ungaran      good    No Hair      blue        red   
3   Male  Human / Radiation       bad    No Hair     green    Unknown   
4   Male              Human       bad    No Hair      blue    Unknown   

            Hero          Publisher  Height  Weight  ...  Weather Control  \
0         A-Bomb      Marvel Comics   203.0   441.0  ...            False   
1     Abe Sapien  Dark Horse Comics   191.0    65.0  ...            False   
2       Abin Sur          DC Comics   185.0    90.0  ...            False   
3    Abomination      Marvel Comics   203.0   441.0  ...            False   
4  Absorbing Man      Marvel Comics   193.0   122.0  ...            False   

   Omnipresent  Omniscient  Hair Manipulation  Nova Force  Odin Force  \
0        False       False              False       False       False   
1        False       False              False       False       False   
2        False       False              False       False       False   
3        False       False              False       False       False   
4        False       False              False       False       False   

   Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0          False               False    False           False  
1          False               False    False           False  
2          False               False    False           False  
3          False               False    False           False  
4          False               False    False           False  

[5 rows x 177 columns]

### 1. Compare the average weight of superheroes who have Super Speed to those who do not.

In [155]:
hero_df.groupby('Super Speed')['Weight'].mean().round(2)

Super Speed
False    101.77
True     129.40
Name: Weight, dtype: float64

The average weight of superheroes who have Super Speed is 129.40 kg, and the average weight of those who do not is 101.77 kg.

### 2. What is the average height of heroes for each publisher?

In [156]:
hero_df.groupby('Publisher')['Height'].mean().round(2)

Publisher
DC Comics            181.92
Dark Horse Comics    176.91
George Lucas         159.60
Image Comics         211.00
Marvel Comics        191.55
Shueisha             171.50
Star Trek            181.50
Team Epic TV         180.75
Unknown              178.00
Name: Height, dtype: float64

In above we have average heights of heros for each publisher in centimeters